internal/providers/gitlab/pull_request_properties.go

```diff
@@ -0,0 +1,186 @@
+//
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitlab
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/xanzy/go-gitlab"
+
+	"github.com/stacklok/minder/internal/entities/properties"
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+	provifv1 "github.com/stacklok/minder/pkg/providers/v1"
+)
+
+func (c *gitlabClient) getPropertiesForPullRequest(
+	ctx context.Context, getByProps *properties.Properties,
+) (*properties.Properties, error) {
+	uid, err := getByProps.GetProperty(properties.PropertyUpstreamID).AsString()
+	if err != nil {
+		return nil, fmt.Errorf("upstream ID not found or invalid: %w", err)
+	}
+
+	iid, err := getByProps.GetProperty(PullRequestNumber).AsString()
+	if err != nil {
+		return nil, fmt.Errorf("merge request number not found or invalid: %w", err)
+	}
+
+	pid, err := getByProps.GetProperty(PullRequestProjectID).AsString()
+	if err != nil {
+		return nil, fmt.Errorf("project ID not found or invalid: %w", err)
+	}
+
+	mrURLPath, err := url.JoinPath("projects", pid, "merge_requests", iid)
+	if err != nil {
+		return nil, fmt.Errorf("failed to join URL path for merge request using upstream ID: %w", err)
+	}
+
+	// NOTE: We're not using github.com/xanzy/go-gitlab to do the actual
+	// request here because of the way they form authentication for requests.
+	// It would be ideal to use it, so we should consider contributing and making
+	// that part more pluggable.
+	req, err := c.NewRequest("GET", mrURLPath, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+
+	resp, err := c.Do(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get merge request: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		if resp.StatusCode == http.StatusNotFound {
+			return nil, provifv1.ErrEntityNotFound
+		}
+
+		return nil, fmt.Errorf("failed to get merge request: %s", resp.Status)
+	}
+
+	mr := &gitlab.MergeRequest{}
+	if err := json.NewDecoder(resp.Body).Decode(mr); err != nil {
+		return nil, fmt.Errorf("failed to decode response: %w", err)
+	}
+
+	// Validate - merge request upstream ID must match the one we requested
+	if FormatPullRequestUpstreamID(mr.ID) != uid {
+		return nil, fmt.Errorf("merge request ID mismatch: %s != %s", FormatPullRequestUpstreamID(mr.ID), uid)
+	}
```

Comment:
**nit:** this might be a bit more readable, feel free to ignore.

```suggestion
	// Validate - merge request upstream ID must match the one we requested
	if res := FormatPullRequestUpstreamID(mr.ID); res != uid {
		return nil, fmt.Errorf("merge request ID mismatch: %s != %s", res, uid)
	}
```